src/components/views/dialogs/security/AccessSecretStorageDialog.tsx

```diff
@@ -201,6 +208,50 @@ export default class AccessSecretStorageDialog extends React.PureComponent<IProp
         });
     };
 
+    private onResetAllClick = (ev: React.MouseEvent<HTMLAnchorElement>) => {
+        ev.preventDefault();
+        this.setState({resetting: true});
+    };
+
+    private onConfirmResetAllClick = async () => {
+        // Hide ourselves so the user can interact with the reset dialogs.
+        // We don't conclude the promise chain (onFinished) yet to avoid confusing
+        // any upstream code flows.
+        //
+        // Note: this will unmount us, so don't call `setState` or anything in the
+        // rest of this function.
+        Modal.toggleCurrentDialogVisibility();
+
+        // Force reset secret storage (which resets the key backup)
+        await accessSecretStorage(async () => {
+            // Now reset cross-signing so everything Just Works™ again.
+            const cli = MatrixClientPeg.get();
+            await cli.bootstrapCrossSigning({
+                authUploadDeviceSigningKeys: async (makeRequest) => {
+                    // XXX: Making this an import breaks the app.
+                    const InteractiveAuthDialog = sdk.getComponent("views.dialogs.InteractiveAuthDialog");
+                    const { finished } = Modal.createTrackedDialog(
+                        'Cross-signing keys dialog', '', InteractiveAuthDialog,
+                        {
+                            title: _t("Setting up keys"),
+                            matrixClient: cli,
+                            makeRequest,
+                        },
+                    );
+                    const [confirmed] = await finished;
+                    if (!confirmed) {
+                        throw new Error("Cross-signing key upload auth canceled");
+                    }
+                },
+                setupNewCrossSigning: true,
+            });
+
+            // Now we can indicate that the user is done pressing buttons, finally.
+            // Upstream flows will detect the new secret storage, key backup, etc and use it.
+            this.props.onFinished(true);
```

Comment:
What happens if something above throws an error...? We won't get here, so will the UI appear to hang?

Reply:
almost certainly, yea. Would it be okay to do an `onFinished(false)` in a catch?

Not sure how easily we can recover from errors here.

Reply:
Yeah, I think that's a good place to start at least, it should be enough to percolate up to higher level paths.